aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -68,7 +69,99 @@
 #include <numeric>
 #include <vector>
 
-namespace at { namespace native {
+namespace at {
+namespace meta {
+
+native::SCATTER_GATHER_OP get_operator_enum(const c10::string_view reduce) {
+  if (reduce == "add") {
+    return native::SCATTER_GATHER_OP::REDUCE_ADD;
+  } else if (reduce == "multiply") {
+    return native::SCATTER_GATHER_OP::REDUCE_MULTIPLY;
+  } else {
+    TORCH_CHECK(false, "reduce argument must be either add or multiply.");
+  }
+}
+
+template <typename Meta>
+void scatter_meta_impl(
+    Meta& meta,
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const c10::optional<Tensor>& src = nullopt,
+    const c10::optional<c10::string_view> reduce = nullopt) {
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, self.dim());
+  at::native::scatter_gather_dtype_check("scatter", self, index, src);
+  at::native::scatter_shape_check(self, wrapped_dim, index, src);
+  auto output = meta.maybe_get_output(0);
+
+  if (output.defined()) {
+    at::assert_no_internal_overlap(output);
+    at::assert_no_overlap(output, index);
+    if (src.has_value()) {
+      at::assert_no_overlap(output, src.value());
+    }
+  }
+
+  meta.set_output(self.sizes(), self.options());
+  if (reduce.has_value()) {
+    auto op = get_operator_enum(reduce.value());
+
+    // On CUDA, if reduce='multiply', we only allow floating point types. That
+    // is because atomic multiplication on CUDA is only implemented for them in
+    // aten/src/THC/THCAtomics.cuh
+    if (self.device().is_cuda() &&
+        op == native::SCATTER_GATHER_OP::REDUCE_MULTIPLY) {
+      TORCH_CHECK(
+          at::isFloatingType(self.scalar_type()),
+          "scatter(): Expected floating type for self when reduce=multiply.");
+    }
+
+    // No bfloat16 dispatch for the scatter-gather kernel on CPU
+    if (self.device() == kCPU) {
+      TORCH_CHECK(
+          self.scalar_type() != kBFloat16,
+          "scatter(): bfloat16 not supported on cpu")
+    }
```

Comment:
This works, but it might be better to put the CPU/CUDA specific checks in their own kernels (because you can think of it as a backend specific implementation detail, as opposed to something that is backend generic)

Reply:
That does make sense. Since the kernels are using `AT_DISPATCH_ALL_TYPES...` (which will error out on not supported types), I guess I could simply remove these checks in the meta function. What do you think?

Reply:
Yeah, that's what I would do in that case